src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+std::shared_ptr<ExpressionResult<T>> ExpressionEvaluator::_evaluate_is_null_expression(
+    const IsNullExpression& expression) {
+  Fail("Can only evaluate predicates to int32_t (aka bool)");
+}
+
+template <>
+std::shared_ptr<ExpressionResult<int32_t>> ExpressionEvaluator::_evaluate_in_expression<int32_t>(
+    const InExpression& in_expression) {
+  const auto& left_expression = *in_expression.value();
+  const auto& right_expression = *in_expression.set();
+
+  std::vector<int32_t> result_values;
+  std::vector<bool> result_nulls;
+
+  if (right_expression.type == ExpressionType::List) {
+    const auto& array_expression = static_cast<const ListExpression&>(right_expression);
+
+    /**
+     * To keep the code simple for now, transform the InExpression like this:
+     * "a IN (x, y, z)"   ---->   "a = x OR a = y OR a = z"
+     *
+     * But first, out of array_expression.elements(), pick those expressions whose type can be compared with
+     * in_expression.value() so we're not getting "Can't compare Int and String" when doing something crazy like
+     * "5 IN (6, 5, "Hello")
+     */
+    const auto left_is_string = left_expression.data_type() == DataType::String;
+    std::vector<std::shared_ptr<AbstractExpression>> type_compatible_elements;
+    for (const auto& element : array_expression.elements()) {
+      if ((element->data_type() == DataType::String) == left_is_string) {
+        type_compatible_elements.emplace_back(element);
+      }
+    }
+
+    if (type_compatible_elements.empty()) {
+      // `5 IN ()` is FALSE as is `NULL IN ()`
+      return std::make_shared<ExpressionResult<int32_t>>(std::vector<int32_t>{0});
+    }
+
+    std::shared_ptr<AbstractExpression> predicate_disjunction =
+        equals(in_expression.value(), type_compatible_elements.front());
+    for (auto element_idx = size_t{1}; element_idx < type_compatible_elements.size(); ++element_idx) {
+      const auto equals_element = equals(in_expression.value(), type_compatible_elements[element_idx]);
+      predicate_disjunction = or_(predicate_disjunction, equals_element);
+    }
+
+    return evaluate_expression_to_result<int32_t>(*predicate_disjunction);
+
+  } else if (right_expression.type == ExpressionType::Select) {
+    const auto* select_expression = dynamic_cast<const PQPSelectExpression*>(&right_expression);
+    Assert(select_expression, "Expected PQPSelectExpression");
+
+    resolve_data_type(select_expression->data_type(), [&](const auto select_data_type_t) {
+      using SelectDataType = typename decltype(select_data_type_t)::type;
+
+      const auto select_result_tables = _evaluate_select_expression_to_tables(*select_expression);
+      const auto select_result_columns = _prune_tables_to_expression_results<SelectDataType>(select_result_tables);
+
+      Assert(select_result_columns.size() == 1 || select_result_columns.size() == _output_row_count,
+             "Unexpected number of lists returned from Select. Should be one, or one per row");
```

Comment:
I don't understand "one per row".

Reply:
Ah, this seems to have something to do with correlation. I think we should explain that more carefully.

Reply:
Changed that to 

```c++
      Assert(select_result_columns.size() == 1 || select_result_columns.size() == _output_row_count,
             "Unexpected number of lists returned from Select. "
             "Should be one (if the Select is uncorrelated), or one per row (if it is)");
```